src/Xamarin.Android.Tools.ApiXmlAdjuster/JavaApi.XmlModel.cs

```diff
@@ -24,15 +26,61 @@ public JavaPackage (JavaApi? parent)
 		{
 			Parent = parent;
 			
-			Types = new List<JavaType> ();
+			Types = new Dictionary<string, List<JavaType>> (StringComparer.OrdinalIgnoreCase);
 		}
 		
 		public  JavaApi?                Parent              { get; private set; }
 
 		public  string?                 Name                { get; set; }
 		public  string?                 JniName             { get; set; }
-		public  IList<JavaType>         Types               { get; set; }
-		
+
+		// Yes, there can be multiple types with the same *Java* name.
+		// For example:
+		// - MyInterface
+		// - MyInterfaceConsts
+		// It's debatable whether we handle this "properly", as most callers just
+		// do `First ()`, but it's been working for years so I'm not changing it.
+		// Exposes an IReadOnlyDictionary so caller cannot bypass our AddType/RemoveType code.
+		public  IReadOnlyDictionary<string, List<JavaType>>   Types               { get; }
+
+		// Use this for a flat list of *all* types
+		public IEnumerable<JavaType> AllTypes => Types.Values.SelectMany (v => v);
+
+		public void AddType (string name, JavaType type)
+		{
+			// If this is a duplicate key, add it to existing list
+			if (Types.TryGetValue (name, out var list)) {
+				list.Add (type);
+				return;
+			}
+
+			// Add to a new list
+			var new_list = new List<JavaType> ();
+			new_list.Add (type);
```

Comment:
Should this check for existing duplicates, so that `AddType(foo); AddType(foo);` only adds `foo` *once*?

Reply:
Maybe? Without full knowledge of every consumer of this class, I opted to not change any *semantics* about it.  The existing code does not make that check, so I left it as is.

Reply:
Fair.